File: boat-core/src/test/java/test/java/xyz/srclab/common/exception/ExceptionTest.java
package test.java.xyz.srclab.common.exception;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.testng.Assert;
import org.testng.annotations.Test;
import xyz.srclab.common.exception.ImpossibleException;
import xyz.srclab.common.exception.StatusException;
import xyz.srclab.common.status.Status;
import xyz.srclab.common.status.StringStatus;

/**
 * @author sunqian
 */
public class ExceptionTest {

    @Test
    public void test() {
        ImpossibleException impossibleException = new ImpossibleException();
        TestException testException = new TestException(StatusException.INTERNAL_STATUS, impossibleException);
        Assert.assertEquals(testException.getDescription(), StatusException.INTERNAL_STATUS.getDescription());
        Assert.assertEquals(testException.getCause(), impossibleException);

        TestException testException2 = new TestException("8", "888", impossibleException);
        Status<String, String, StringStatus> exceptionStatus = testException2.withMoreDescription("999");
        Assert.assertEquals(exceptionStatus.getDescription(), "888[999]");
        Status<String, String, StringStatus> exceptionStatus2 = testException2.withNewDescription("999");
        Assert.assertEquals(exceptionStatus2.getDescription(), "999");
    }

    public static class TestException extends StatusException {

        public TestException(@NotNull Status<String, String, StringStatus> exceptionStatus, @Nullable Throwable cause) {
            super(exceptionStatus, cause);
        }

        public TestException(@NotNull String code, @Nullable String description, @Nullable Throwable cause) {
            super(code, description, cause);
        }

        @Override
        public String getCode() {
            return super.getCode();
        }

        @Nullable
        @Override
        public String getDescription() {
            return super.getDescription();
        }
    }
}

File: srclab-common/src/test/kotlin/test/xyz/srclab/common/exception/ExceptionTest.kt
package test.xyz.srclab.common.exception

import org.testng.annotations.Test
import xyz.srclab.common.exception.CommonException
import xyz.srclab.common.exception.DefaultExceptionStatus
import xyz.srclab.common.exception.ExceptionStatus
import xyz.srclab.common.exception.RuntimeExceptionWrapper
import xyz.srclab.common.state.StateKit
import xyz.srclab.test.doAssertEquals

/**
 * @author sunqian
 */
object ExceptionTest {

    @Test
    fun testBusinessException() {
        val cause = IllegalStateException()
        val b1 = CommonException(DefaultExceptionStatus.INTERNAL)
        doAssertEquals(b1.code, DefaultExceptionStatus.INTERNAL.code)
        doAssertEquals(b1.description, DefaultExceptionStatus.INTERNAL.description)
        doAssertEquals(StateKit.equals(b1, DefaultExceptionStatus.INTERNAL), true)
        doAssertEquals(b1 == b1, true)
        doAssertEquals(
            b1.withMoreDescription("sss"),
            DefaultExceptionStatus.INTERNAL.withMoreDescription("sss")
        )
        doAssertEquals(b1 == cause, false)

        val status = ExceptionStatus.from(DefaultExceptionStatus.INTERNAL)
        doAssertEquals(status, b1)
        doAssertEquals(status.hashCode(), b1.hashCode())
        doAssertEquals(
            status.withMoreDescription("sss"),
            b1.withMoreDescription("sss")
        )

        doAssertEquals(ExceptionStatus.newExceptionStatus("666").code, "666")
        doAssertEquals(ExceptionStatus.newExceptionStatus("666").description, null)
    }

    @Test
    fun testExceptionWrapper() {
        val cause = IllegalStateException()
        val wrapper = RuntimeExceptionWrapper(cause)
        doAssertEquals(wrapper.wrapped, cause)
    }
}